### PR TITLE
Intergrate nmpolicy

### PR DESCRIPTION
**Pull request by AlonaKaplan**
**Is this a BUG FIX or a FEATURE ?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug

/kind enhancement

**What this PR does / why we need it**:
The PR is intoducing `nncp.spec.capture` which can include a map (string -> string) of captureIDs to expressions that can be referenced at desiredState and at another capture.

The captures and the desired state references are resolved by invoking `nmpolicy.GenerateState` (NMPolicy - https://github.com/nmstate/nmpolicy).

The resolved captures cache is stored on the enactment (nnce.status.nmpolicyStatus.cache.capture) and passed to the `nmpolicy.GenerateState` in case the nncp is updated.

**Special notes for your reviewer**:
The PR doesn't take care of input validation and reconcile corner cases (reboot of the node while the policy is applied, etc...).
A follow-up PRs will take of of these.

**Release note**:

```release-note
Allows adding captures to node network configuration policy spec
```

### CODE REVIEW

File: api/shared/nodenetworkconfigurationenactment_types.go [hunk 1]
```diff
@@ -13,10 +14,35 @@ type NodeNetworkConfigurationEnactmentStatus struct {
 	// the policy desiredState as template
 	DesiredState State `json:"desiredState,omitempty"`
 
+	// CapturedState The captured state from NMPolicy
+	// after processing the capture at NNCP
+	NMPolicyStatus NodeNetworkConfigurationEnactmentNMPolicyStatus `json:"nmpolicyStatus,omitempty"`
```
Comment (phoracek) [inline api/shared/nodenetworkconfigurationenactment_types.go:19]: I have mixed feeling about the name. We don't mention nmstate in our API, I think we should not mention nmpolicy neither. It also makes it confusing as there is nmpolicy as a part of policy. Could we grant it a more generic name? "templating" or something similar.

Calling in the namemaster @EdDev

Comment (AlonaKaplan) [reply api/shared/nodenetworkconfigurationenactment_types.go:19]: Maybe `CaptureStatus`?

File: api/shared/nodenetworkconfigurationenactment_types.go [hunk 2]
```diff
@@ -13,10 +14,34 @@ type NodeNetworkConfigurationEnactmentStatus struct {
...
 	// The generation from policy needed to check if an enactment
 	// condition status belongs to the same policy version
-	PolicyGeneration int64         `json:"policyGeneration,omitempty"`
-	Conditions       ConditionList `json:"conditions,omitempty"`
+	PolicyGeneration int64 `json:"policyGeneration,omitempty"`
+
+	Conditions ConditionList `json:"conditions,omitempty"`
+}
+
+type NodeNetworkConfigurationEnactmentCaptureStatus struct {
```
Comment (EdDev) [inline api/shared/nodenetworkconfigurationenactment_types.go:27]: The way it is structured now, it looks like the name should specify the "cache" and not the capture specifically.

In `nmpolicy`, "cache" was a generalization of the input, which consists only of captures.
So we can either keep the generalization here as well, and call this CachedState, and in it specify the captures, or just do not generalize it and leave only one "level".

BTW, we are talking about "state" and not "status", right?

Comment (AlonaKaplan) [reply api/shared/nodenetworkconfigurationenactment_types.go:27]: It is not just the cache, The field contain both the capture cache and the metainfo of the last call to the `nmpolicy.GenerateState`.
I didn't want to use the term state since we use it in the enactment to represent yamls (desiredState, capture.state).

File: api/shared/nodenetworkconfigurationenactment_types.go (hunk 1 above)
Comment (phoracek) [reply api/shared/nodenetworkconfigurationenactment_types.go]: Capture sounds good to me, thanks o/

File: test/e2e/handler/default_bridged_network_with_nmpolicy_test.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	nmstate "github.com/nmstate/kubernetes-nmstate/api/shared"
+)
+
+var _ = Describe("NodeNetworkConfigurationPolicy default bridged network with nmpolicy", func() {
```
Comment (qinqon) [inline test/e2e/handler/default_bridged_network_with_nmpolicy_test.go:13]: I see too much duplication here, maybe it's better to just parameterize the test with desiredState so we use capture and non capture version, also in the future we will add a non DHCP default bridge here on the `default_bridged_network_test.go`

**Review by qinqon — CHANGES_REQUESTED** [review]
I think we can flat out what we put at Enactment and store top level MetaInfo directly at the enactment and remove the Cache level, also about e2e tests paramterizing the already existing default bridge would be better so we don't duplicate in the future we will do a specific non DHCP version of default bridge that will just use nmpolicy.

File: api/shared/nodenetworkconfigurationenactment_types.go [hunk 4]
```diff
@@ -13,10 +14,34 @@ type NodeNetworkConfigurationEnactmentStatus struct {
 	// the policy desiredState as template
 	DesiredState State `json:"desiredState,omitempty"`
 
+	// CaptureStatus The capture status after processing the capture at NNCP
+	CaptureStatus NodeNetworkConfigurationEnactmentCaptureStatus `json:"captureStatus,omitempty"`
+
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply api/shared/nodenetworkconfigurationenactment_types.go]: done

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 5]
```diff
@@ -288,9 +289,18 @@ func (r *NodeNetworkConfigurationPolicyReconciler) initializeEnactment(policy nm
 		enactmentConditions := enactmentconditions.New(r.APIClient, enactmentKey)
 		enactmentConditions.Reset()
 	}
+	nns, err := r.readNNS(nodeName)
+	if err != nil {
+		return nil, err
+	}
+	captureStatus, desiredState, err := nmpolicy.GenerateState(desiredStateWithDefaults, policy.Spec, nns.Status.CurrentState, enactment.Status.CaptureStatus.Cache)
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go]: The return value was changed anyway to return the `capturedStates` and `desiredStateMetaInfo`.

File: api/shared/nodenetworkconfigurationenactment_types.go [hunk 6]
```diff
@@ -13,10 +14,34 @@ type NodeNetworkConfigurationEnactmentStatus struct {
...
+type NodeNetworkConfigurationEnactmentCaptureStatus struct {
+	Cache    NodeNetworkConfigurationEnactmentCachedState `json:"cache,omitempty"`
+	MetaInfo NodeNetworkConfigurationEnactmentMetaInfo    `json:"metaInfo,omitempty"`
+}
+
+type NodeNetworkConfigurationEnactmentCachedState struct {
+	Capture map[string]NodeNetworkConfigurationEnactmentCaptureState `json:"capture,omitempty"`
+}
+
+type NodeNetworkConfigurationEnactmentCaptureState struct {
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply api/shared/nodenetworkconfigurationenactment_types.go]: done

File: api/shared/nodenetworkconfigurationpolicy_types.go [hunk 7]
```diff
@@ -13,6 +13,10 @@ type NodeNetworkConfigurationPolicySpec struct {
 	// +optional
 	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
 
+	// Capture contains expressions with an associated name than can be referenced
+	// at the DesiredState.
+	Capture map[string]string `json:"capture,omitempty"`
+
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply api/shared/nodenetworkconfigurationpolicy_types.go]: done

File: pkg/nmpolicy/generate_suite_test.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package nmpolicy
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/nmpolicy/generate_suite_test.go]: done

File: pkg/nmpolicy/generate_suite_test.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/onsi/ginkgo/reporters"
+)
+
+func TestUnit(t *testing.T) {
+	RegisterFailHandler(Fail)
+	junitReporter := reporters.NewJUnitReporter("junit.generate_suite_test.xml")
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/nmpolicy/generate_suite_test.go]: done

File: pkg/nmpolicy/generate_suite_test.go [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+	RunSpecsWithDefaultAndCustomReporters(t, "Generate Test Suite", []Reporter{junitReporter})
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/nmpolicy/generate_suite_test.go]: done

File: pkg/nmpolicy/generate_test.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+	nmstateapi "github.com/nmstate/kubernetes-nmstate/api/shared"
+	nmpolicytypes "github.com/nmstate/nmpolicy/nmpolicy/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Generate nmpolicy", func() {
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/nmpolicy/generate_test.go]: done

File: pkg/nmpolicy/generate_test.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+	When("nmpolicy generate fails", func() {
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/nmpolicy/generate_test.go]: done

File: pkg/nmpolicy/generate_test.go [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+				nmstateapi.State{},
+				nmstateapi.NodeNetworkConfigurationEnactmentCachedState{})
+			Expect(err).To(HaveOccurred())
+			Expect(captureStatus).To(Equal(nmstateapi.NodeNetworkConfigurationEnactmentCaptureStatus{}))
+			Expect(desiredState).To(Equal(nmstateapi.State{}))
+		})
+
+	})
+
+	When("When nmpolicy generate succeeds", func() {
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/nmpolicy/generate_test.go]: done

File: pkg/nmpolicy/generate_test.go [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+	shouldFail bool
+	output     nmpolicytypes.GeneratedState
+}
+
+func (n nmpolicyStub) GenerateState(nmpolicySpec nmpolicytypes.PolicySpec, currentState []byte, cache nmpolicytypes.CachedState) (nmpolicytypes.GeneratedState, error) {
+	if n.shouldFail {
+		return nmpolicytypes.GeneratedState{}, fmt.Errorf("GenerateStateFailed")
+	}
+	return n.output, nil
+}
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/nmpolicy/generate_test.go:86]: I don't understand the comment.

File: test/e2e/handler/default_bridged_network_with_nmpolicy_test.go (hunk 3 above)
Comment (AlonaKaplan) [reply test/e2e/handler/default_bridged_network_with_nmpolicy_test.go:13]: As we discussed offline, the test avoid duplication as much as possible and follows the principals of the other handler e2e tests.

File: test/e2e/handler/default_bridged_network_with_nmpolicy_test.go [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+    state: up
+    ipv4:
+      dhcp: true
+      enabled: true
+    bridge:
+      options:
+        stp:
+          enabled: false
+      port:
+      - name: "{{ capture.base-iface.interfaces.0.name }}"
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply test/e2e/handler/default_bridged_network_with_nmpolicy_test.go]: done.

**Comment by qinqon** [comment]
@rhrazdil can you take a look ?

**Comment by qinqon** [comment]
@AlonaKaplan can you pin to the new NMPolicy release https://github.com/nmstate/nmpolicy/releases/tag/v0.1.0-alpha.1

File: test/e2e/handler/utils.go [hunk 16]
```diff
@@ -68,12 +68,13 @@ func interfaceByName(interfaces []interface{}, searchedName string) map[string]i
 	return dummy
 }
 
-func setDesiredStateWithPolicyAndNodeSelector(name string, desiredState nmstate.State, nodeSelector map[string]string) error {
+func setDesiredStateWithPolicyAndCapturesAndNodeSelector(name string, desiredState nmstate.State, capture map[string]string, nodeSelector map[string]string) error {
```
Comment (rhrazdil) [inline test/e2e/handler/utils.go:71]: s/Captures/Capture

**Review by qinqon — CHANGES_REQUESTED** [review]

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 17]
```diff
@@ -289,10 +291,37 @@ func (r *NodeNetworkConfigurationPolicyReconciler) initializeEnactment(policy nm
...
+	}
+
+	desiredStateWithDefaults, err := nmstate.ApplyDefaultVlanFiltering(policy.Spec.DesiredState)
+	if err != nil {
+		return false, errors.Wrap(err, "error applying defaults to policy desiredState")
+	}
+
+	capturedStates, desiredStateMetaInfo, desiredState, err := nmpolicy.GenerateState(desiredStateWithDefaults, policy.Spec, nns.Status.CurrentState, enactment.Status.CapturedStates)
+	if err != nil {
+		enactmentstatus.Update(r.APIClient, nmstateapi.EnactmentKey(nodeName, policy.Name), func(status *nmstateapi.NodeNetworkConfigurationEnactmentStatus) {
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go:310]: What do you mean? To check what about the error?

[the remark this answers is not in the document]
Comment (qinqon) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go:310]: `Update` function is returning an error that we are ignoring, https://github.com/nmstate/kubernetes-nmstate/blob/main/pkg/enactmentstatus/status.go#L26

File: test/e2e/handler/utils.go (hunk 16 above)
Comment (AlonaKaplan) [reply test/e2e/handler/utils.go]: done

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 18]
```diff
@@ -156,20 +157,26 @@ func (r *NodeNetworkConfigurationPolicyReconciler) Reconcile(ctx context.Context
...
 	}
 
-	enactmentInstance, err := r.enactmentForPolicy(instance)
+	enactmentConditions := enactmentconditions.New(r.APIClient, nmstateapi.EnactmentKey(nodeName, instance.Name))
+
+	stopReconcile, err := r.updateEnactmentDesiredState(*instance, *enactmentInstance, enactmentConditions)
+	if err != nil {
+		return ctrl.Result{Requeue: !stopReconcile}, err
+	}
+
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go]: done.

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 19]
```diff
@@ -289,10 +291,37 @@ func (r *NodeNetworkConfigurationPolicyReconciler) initializeEnactment(policy nm
 		enactmentConditions.Reset()
 	}
 
-	return &enactment.Status.Conditions, enactmentstatus.Update(r.APIClient, enactmentKey, func(status *nmstateapi.NodeNetworkConfigurationEnactmentStatus) {
-		status.DesiredState = desiredStateWithDefaults
+	return &enactment, nil
+}
+
+func (r *NodeNetworkConfigurationPolicyReconciler) updateEnactmentDesiredState(policy nmstatev1.NodeNetworkConfigurationPolicy, enactment nmstatev1beta1.NodeNetworkConfigurationEnactment, enactmentConditions enactmentconditions.EnactmentConditions) (bool /* stop reconcile*/, error) {
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go]: done.

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 20]
```diff
@@ -289,10 +291,37 @@ func (r *NodeNetworkConfigurationPolicyReconciler) initializeEnactment(policy nm
...
+	nns, err := r.readNNS(nodeName)
+	if err != nil {
+		return false, err
+	}
+
+	desiredStateWithDefaults, err := nmstate.ApplyDefaultVlanFiltering(policy.Spec.DesiredState)
+	if err != nil {
+		return false, errors.Wrap(err, "error applying defaults to policy desiredState")
+	}
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go]: done.

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 21]
```diff
@@ -289,10 +291,37 @@ func (r *NodeNetworkConfigurationPolicyReconciler) initializeEnactment(policy nm
...
+
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go:310]: done.

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go]: done.

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 22]
```diff
@@ -289,10 +291,37 @@ func (r *NodeNetworkConfigurationPolicyReconciler) initializeEnactment(policy nm
...
+	capturedStates, desiredStateMetaInfo, desiredState, err := nmpolicy.GenerateState(desiredStateWithDefaults, policy.Spec, nns.Status.CurrentState, enactment.Status.CapturedStates)
+	if err != nil {
+		enactmentstatus.Update(r.APIClient, nmstateapi.EnactmentKey(nodeName, policy.Name), func(status *nmstateapi.NodeNetworkConfigurationEnactmentStatus) {
+			status.PolicyGeneration = policy.Generation
+		})
+		enactmentConditions.NotifyGenerateFailure(err)
+		return true /* stop reconcile*/, err
+	}
+
+	enactmentstatus.Update(r.APIClient, nmstateapi.EnactmentKey(nodeName, policy.Name), func(status *nmstateapi.NodeNetworkConfigurationEnactmentStatus) {
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go]: done.

File: pkg/enactmentstatus/conditions/conditions.go [hunk 23]
```diff
@@ -29,12 +29,12 @@ func New(client client.Client, enactmentKey types.NamespacedName) EnactmentCondi
 	return conditions
 }
 
-func (ec *EnactmentConditions) NotifyNodeSelectorFailure(err error) {
-	ec.logger.Info("NotifyNodeSelectorFailure")
-	message := fmt.Sprintf("failure checking node selectors : %v", err)
+func (ec *EnactmentConditions) NotifyGenerateFailure(err error) {
+	ec.logger.Info("NotifyGenerateFailure")
+	message := fmt.Sprintf("failure generating captures : %v", err)
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/enactmentstatus/conditions/conditions.go]: done

**Comment by AlonaKaplan** [comment]
[Comment fixes](https://github.com/nmstate/kubernetes-nmstate/compare/5e72833b9623c0adefdc894d9b533698b6fc9484..7ef8da0d78a1aac6b027ac659ee4bfa7bdaa561f).
[Rebase](https://github.com/nmstate/kubernetes-nmstate/compare/7ef8da0d78a1aac6b027ac659ee4bfa7bdaa561f..e0a5d30ca2fff668d386bb974696f6b03350702f)

**Comment by AlonaKaplan** [comment]
[Fixing small compilation error.](https://github.com/nmstate/kubernetes-nmstate/compare/e0a5d30ca2fff668d386bb974696f6b03350702f..bcc091969a16c6a3435c03a10dfa506367204abb)

**Comment by AlonaKaplan** [comment]
[Fix linter errros.](https://github.com/nmstate/kubernetes-nmstate/compare/bcc091969a16c6a3435c03a10dfa506367204abb..60b26a2352c018f3b1d5c20ea0f2204d7bbe3f74)

@radim the linter didn't like the new line in the beginning of a method.

**Review by qinqon — APPROVED** [review]
/lgtm

File: pkg/nmpolicy/generate.go [hunk 24]
```diff
@@ -9,26 +9,43 @@ import (
...
+}
+
+func GenerateState(desiredState nmstateapi.State,
+	policySpec nmstateapi.NodeNetworkConfigurationPolicySpec,
+	currentState nmstateapi.State,
+	cachedState map[string]nmstateapi.NodeNetworkConfigurationEnactmentCapturedState) (map[string]nmstateapi.NodeNetworkConfigurationEnactmentCapturedState, nmstateapi.NodeNetworkConfigurationEnactmentMetaInfo, nmstateapi.State, error) {
+	return GenerateStateWithStateGenerator(GenerateStateWithNMPolicy{}, desiredState, policySpec, currentState, cachedState)
 }
 
 func convertGeneratedStateToEnactment(nmpolicyGeneratedState nmpolicytypes.GeneratedState) (map[string]nmstateapi.NodeNetworkConfigurationEnactmentCapturedState, nmstateapi.NodeNetworkConfigurationEnactmentMetaInfo, nmstateapi.State) {
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply pkg/nmpolicy/generate.go]: Yes, thanks

[the remark this answers is not in the document]
Comment (rhrazdil) [reply pkg/nmpolicy/generate.go]: `convertGeneratedStateToEnactmentConfig` sounds better, let's go with that

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 25]
```diff
@@ -289,21 +290,47 @@ func (r *NodeNetworkConfigurationPolicyReconciler) initializeEnactment(policy nm
...
+	capturedStates, desiredStateMetaInfo, generatedDesiredState, err := nmpolicy.GenerateState(policy.Spec.DesiredState, policy.Spec, nns.Status.CurrentState, enactment.Status.CapturedStates)
 	if err != nil {
-		return nil, err
+		err = enactmentstatus.Update(r.APIClient, nmstateapi.EnactmentKey(nodeName, policy.Name), func(status *nmstateapi.NodeNetworkConfigurationEnactmentStatus) {
+			status.PolicyGeneration = policy.Generation
+		})
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		enactmentConditions.NotifyGenerateFailure(err)
```
Comment (rhrazdil) [inline controllers/handler/nodenetworkconfigurationpolicy_controller.go:313]: `err` is `nil` here, isn't it?

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 26]
```diff
@@ -289,21 +290,47 @@ func (r *NodeNetworkConfigurationPolicyReconciler) initializeEnactment(policy nm
...
+		return ctrl.Result{Requeue: false}, err
 	}
 
-	return &enactment.Status.Conditions, enactmentstatus.Update(r.APIClient, enactmentKey, func(status *nmstateapi.NodeNetworkConfigurationEnactmentStatus) {
-		status.DesiredState = desiredState
+	desiredStateWithDefaults, err := nmstate.ApplyDefaultVlanFiltering(generatedDesiredState)
+	if err != nil {
+		return ctrl.Result{Requeue: false}, errors.Wrap(err, "error applying defaults to policy desiredState")
```
Comment (rhrazdil) [inline controllers/handler/nodenetworkconfigurationpolicy_controller.go:319]: This is tricky, we can't return an error here as the Requeue value is ignored. See the reque logic in controller-runtime: https://github.com/kubernetes-sigs/controller-runtime/blob/8f43f26d53ffb08a96ffce1298df0cd761f8be0d/pkg/internal/controller/controller.go#L311

If err != nil , the Request is requeued regardles of the Result.

Can we just log the error?

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 27]
```diff
@@ -289,21 +290,47 @@ func (r *NodeNetworkConfigurationPolicyReconciler) initializeEnactment(policy nm
...
 	if err != nil {
-		return nil, err
+		err = enactmentstatus.Update(r.APIClient, nmstateapi.EnactmentKey(nodeName, policy.Name), func(status *nmstateapi.NodeNetworkConfigurationEnactmentStatus) {
+			status.PolicyGeneration = policy.Generation
+		})
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		enactmentConditions.NotifyGenerateFailure(err)
+		return ctrl.Result{Requeue: false}, err
```
Comment (rhrazdil) [inline controllers/handler/nodenetworkconfigurationpolicy_controller.go:314]: `err` -> `nil`

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 28]
```diff
@@ -289,21 +290,47 @@ func (r *NodeNetworkConfigurationPolicyReconciler) initializeEnactment(policy nm
...
 	}
 
-	return &enactment.Status.Conditions, enactmentstatus.Update(r.APIClient, enactmentKey, func(status *nmstateapi.NodeNetworkConfigurationEnactmentStatus) {
-		status.DesiredState = desiredState
+	desiredStateWithDefaults, err := nmstate.ApplyDefaultVlanFiltering(generatedDesiredState)
+	if err != nil {
+		return ctrl.Result{Requeue: false}, errors.Wrap(err, "error applying defaults to policy desiredState")
+	}
+
+	err = enactmentstatus.Update(r.APIClient, nmstateapi.EnactmentKey(nodeName, policy.Name), func(status *nmstateapi.NodeNetworkConfigurationEnactmentStatus) {
```
Comment (rhrazdil) [inline controllers/handler/nodenetworkconfigurationpolicy_controller.go:322]: Why don't apply changes and `Update` only once?

File: pkg/enactmentstatus/conditions/conditions.go [hunk 29]
```diff
@@ -29,12 +29,12 @@ func New(client client.Client, enactmentKey types.NamespacedName) EnactmentCondi
...
-func (ec *EnactmentConditions) NotifyNodeSelectorFailure(err error) {
-	ec.logger.Info("NotifyNodeSelectorFailure")
-	message := fmt.Sprintf("failure checking node selectors : %v", err)
+func (ec *EnactmentConditions) NotifyGenerateFailure(err error) {
+	ec.logger.Info("NotifyGenerateFailure")
+	message := fmt.Sprintf("failure generating desiredState and capturedStates: %v", err)
 	err = ec.updateEnactmentConditions(SetFailedToConfigure, message)
 	if err != nil {
-		ec.logger.Error(err, "Error notifying state NodeSelectorNotMatching with failure")
+		ec.logger.Error(err, "Error notifying state generate captures with failure")
```
Comment (rhrazdil) [inline pkg/enactmentstatus/conditions/conditions.go:37]: It's not really part of your PR, but could you please add the err content to the message?

**Review by rhrazdil — CHANGES_REQUESTED** [review]
Really nice, I only have one important find, the rest are minor things/questions

**Comment by qinqon** [comment]
@AlonaKaplan we need to re-pin nmpolicy to https://github.com/nmstate/nmpolicy/releases/tag/v0.1.0-alpha.3, it fixes a panic.

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go (hunk 28 above)
Comment (rhrazdil) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go:324]: ah of course... Thanks!

**Review by rhrazdil — APPROVED** [review]
/lgtm

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go [hunk 30]
```diff
@@ -289,10 +291,48 @@ func (r *NodeNetworkConfigurationPolicyReconciler) initializeEnactment(policy nm
 		enactmentConditions.Reset()
 	}
 
-	return &enactment.Status.Conditions, enactmentstatus.Update(r.APIClient, enactmentKey, func(status *nmstateapi.NodeNetworkConfigurationEnactmentStatus) {
+	return &enactment, nil
+}
+
+func (r *NodeNetworkConfigurationPolicyReconciler) fillInEnactmentStatus(policy nmstatev1.NodeNetworkConfigurationPolicy,
```
Comment (qinqon) [inline controllers/handler/nodenetworkconfigurationpolicy_controller.go:297]: Thinking about this, the Update function is already retrying so it does not make sense to Reconcile if Update fail, so if we just return the error if update or the other stuff fails and we read the NNS outside we can do the following changing the function:

```golang
func (r *NodeNetworkConfigurationPolicyReconciler) fillInEnactmentStatus(nns, nncp, nnce, enactmentConditions) error {...}      
```

And at the caller

```golang

nns, err := r.readNNS(nodeName)
if err != nil {
    ctr.Result{}, err
}

if err := r.fillInEnactmentStatus(nns, *instance, *enactmentInstance, enactmentConditions); err != nil {
    r.Log.Error(err, "failed filling in the NNCE status")
    return ctrl.Result{}, nil
}

```

**Review by qinqon — CHANGES_REQUESTED** [review]
We can simplify the fillIn function, we don't have to Reconcile on Update fail.

File: controllers/handler/nodenetworkconfigurationpolicy_controller.go (hunk 30 above)
Comment (AlonaKaplan) [reply controllers/handler/nodenetworkconfigurationpolicy_controller.go:297]: done.

**Review by qinqon — APPROVED** [review]
/lgtm
/approve
